src/hotspot/share/code/debugInfo.cpp

```diff
@@ -191,6 +216,104 @@ void ObjectValue::print_fields_on(outputStream* st) const {
+// merges in C2. This method will select which path the allocation merge
+// took during execution of the Trap that triggered the rematerialization
+// of the object.
+ObjectValue* ObjectMergeValue::select(frame& fr, RegisterMap& reg_map) {
+  StackValue* sv_selector = StackValue::create_stack_value(&fr, &reg_map, _selector);
+  jint selector = sv_selector->get_int();
+
+  // If the selector is '-1' it means that execution followed the path
+  // where no scalar replacement happened.
+  // Otherwise, it is the index in _possible_objects array that holds
+  // the description of the scalar replaced object.
+  if (selector == -1) {
+    StackValue* sv_merge_pointer = StackValue::create_stack_value(&fr, &reg_map, _merge_pointer);
+    _selected = new ObjectValue(id());
+
+    // Retrieve the pointer to the real object and use it as if we had
+    // allocated it during the deoptimization
+    _selected->set_value(sv_merge_pointer->get_obj()());
+
+    // No need to rematerialize
+    return nullptr;
+  } else {
+    assert(selector < _possible_objects.length(), "sanity");
+    _selected = (ObjectValue*) _possible_objects.at(selector);
+
+    // Set it to true so that the object will get rematerialized
+    if (!_selected->is_root()) {
+      _selected->set_root(true);
+
+      // We can't assume that 'select(...)' will be called before we check if
+      // the candidate needs to be rematerialized or not. Therefore, we need to
+      // return the candidate, now set to 'not only merge candidate', and try to
+      // rematerialize it.
+      return _selected;
+    } else {
+      // Since the object was not only a candidate it will already be rematerialized on its own.
+      return nullptr;
+    }
+  }
+}
+
+void ObjectMergeValue::read_object(DebugInfoReadStream* stream) {
+  _selector = read_from(stream);
+  _merge_pointer = read_from(stream);
+  int ncandidates = stream->read_int();
+  for (int i = 0; i < ncandidates; i++) {
+    ScopeValue* result = read_from(stream);
+    assert(result->is_object(), "Candidate is not an object!");
+    ObjectValue* obj = result->as_ObjectValue();
+    _possible_objects.append(obj);
+  }
+}
+
+void ObjectMergeValue::write_on(DebugInfoWriteStream* stream) {
+  if (is_visited()) {
+    stream->write_int(OBJECT_ID_CODE);
+    stream->write_int(_id);
+  } else {
+    set_visited(true);
+    stream->write_int(OBJECT_MERGE_CODE);
+    stream->write_int(_id);
+    _selector->write_on(stream);
+    _merge_pointer->write_on(stream);
+    int ncandidates = _possible_objects.length();
+    stream->write_int(ncandidates);
+    for (int i = 0; i < ncandidates; i++) {
+      _possible_objects.at(i)->as_ObjectValue()->write_on(stream);
+    }
+  }
+}
+
+void ObjectMergeValue::print_on(outputStream* st) const {
+  st->print("merge: ID=%d", _id);
+}
+
+void ObjectMergeValue::print_detailed(outputStream* st) const {
+  st->print("merge: ID=%d", _id);
+#ifndef PRODUCT
```

Comment:
Can you post a sample of the output, please?

Why is it limited to non-product builds? It's valuable irrespective of build flavor.

As I see in `ObjectValue::print_on` and `ScopeDesc::print_on`, you mix `print_on` with `print_fields_on`. Any particular reason for that? You could add `is_object_merge` case in ObjectValue::print_on` instead and extend `ObjectValue::print_fields_on` to cover `ObjectMergeValue` case. I find it hard to reason about `ObjectValue::print_on` vs `ObjectMergeValue::print_on` since it's a non-virtual method. 

 

Also, formatting is broken.

Reply:
I added a few samples below and there are a few more here: https://gist.github.com/JohnTortugo/913523947e08157def6cfebafa7d5daa

Sample 1:

```
Compiled method (c2)     415   24             TestTrapAfterMerge::test (57 bytes)
 total in heap  [0x00007f7b4d03da90,0x00007f7b4d03de18] = 904
 relocation     [0x00007f7b4d03dc00,0x00007f7b4d03dc18] = 24
 main code      [0x00007f7b4d03dc20,0x00007f7b4d03dcb8] = 152
 stub code      [0x00007f7b4d03dcb8,0x00007f7b4d03dcd0] = 24
 oops           [0x00007f7b4d03dcd0,0x00007f7b4d03dce0] = 16
 metadata       [0x00007f7b4d03dce0,0x00007f7b4d03dce8] = 8
 scopes data    [0x00007f7b4d03dce8,0x00007f7b4d03dd50] = 104
 scopes pcs     [0x00007f7b4d03dd50,0x00007f7b4d03de10] = 192
 dependencies   [0x00007f7b4d03de10,0x00007f7b4d03de18] = 8
scopes:
ScopeDesc(pc=0x00007f7b4d03dc3a offset=1a):
   TestTrapAfterMerge::test@-1 (line 3)
ScopeDesc(pc=0x00007f7b4d03dc41 offset=21):
   TestTrapAfterMerge::test@11 (line 5)
ScopeDesc(pc=0x00007f7b4d03dc44 offset=24):
   TestTrapAfterMerge::test@51 (line 12)
ScopeDesc(pc=0x00007f7b4d03dc4a offset=2a):
   TestTrapAfterMerge::test@46 (line 8)
ScopeDesc(pc=0x00007f7b4d03dc52 offset=32):
   TestTrapAfterMerge::test@37 (line 9)
ScopeDesc(pc=0x00007f7b4d03dc57 offset=37):
   TestTrapAfterMerge::test@43 (line 8)
ScopeDesc(pc=0x00007f7b4d03dc61 offset=41):
   TestTrapAfterMerge::test@46 (line 8)  reexecute=true
   Locals
    - l0: empty
    - l1: empty
    - l2: reg rbx [6],int
    - l3: empty
    - l4: merge: ID=26
    - l5: reg r11 [22],int
   Objects
    - 0: merge: ID=26, selector="reg r10 [20],int", merge_pointer="nullptr", candidate objs=[27, 28]
    - 1: obj: ID=27, is_root=0, N.Fields=1, klass: Point 
Fields: reg r8 [16],int
    - 2: obj: ID=28, is_root=0, N.Fields=1, klass: Point 
Fields: reg rcx [2],int
ScopeDesc(pc=0x00007f7b4d03dc63 offset=43):
   TestTrapAfterMerge::test@46 (line 8)
ScopeDesc(pc=0x00007f7b4d03dc6c offset=4c):
   TestTrapAfterMerge::test@34 (line 8)
ScopeDesc(pc=0x00007f7b4d03dc71 offset=51):
   TestTrapAfterMerge::test@55 (line 12)
```

- Sample2:

```
Compiled method (c2)     443   24             TestManys::test (41 bytes)
 total in heap  [0x00007f35e9155b90,0x00007f35e9155e78] = 744
 relocation     [0x00007f35e9155d00,0x00007f35e9155d18] = 24
 main code      [0x00007f35e9155d20,0x00007f35e9155d88] = 104
 stub code      [0x00007f35e9155d88,0x00007f35e9155da0] = 24
 oops           [0x00007f35e9155da0,0x00007f35e9155db0] = 16
 metadata       [0x00007f35e9155db0,0x00007f35e9155db8] = 8
 scopes data    [0x00007f35e9155db8,0x00007f35e9155e10] = 88
 scopes pcs     [0x00007f35e9155e10,0x00007f35e9155e70] = 96
 dependencies   [0x00007f35e9155e70,0x00007f35e9155e78] = 8
scopes:
ScopeDesc(pc=0x00007f35e9155d3a offset=1a):
   TestManys::test@-1 (line 57)
ScopeDesc(pc=0x00007f35e9155d42 offset=22):
   TestManys::test@11 (line 59)
ScopeDesc(pc=0x00007f35e9155d58 offset=38):
   TestManys::test@25 (line 63)
   Locals
    - l0: empty
    - l1: empty
    - l2: empty
    - l3: empty
    - l4: empty
    - l5: empty
    - l6: empty
    - l7: empty
    - l8: merge: ID=26
   Objects
    - 0: merge: ID=26, selector="reg rbp [10],int", merge_pointer="nullptr", candidate objs=[27, 28]
    - 1: obj: ID=27, is_root=0, N.Fields=4, klass: Point 
Fields: stack[36], stack[36], 0, 0
    - 2: obj: ID=28, is_root=0, N.Fields=4, klass: Point 
Fields: 2023, 0, 0, 0
ScopeDesc(pc=0x00007f35e9155d74 offset=54):
   TestManys::test@25 (line 63)
```

- Sample3:
```
Compiled method (c2)     436   24             TestMultiSFO::test (48 bytes)
 total in heap  [0x00007f1df5155590,0x00007f1df5155850] = 704
 relocation     [0x00007f1df5155700,0x00007f1df5155718] = 24
 main code      [0x00007f1df5155720,0x00007f1df5155788] = 104
 stub code      [0x00007f1df5155788,0x00007f1df51557a0] = 24
 oops           [0x00007f1df51557a0,0x00007f1df51557b0] = 16
 metadata       [0x00007f1df51557b0,0x00007f1df51557b8] = 8
 scopes data    [0x00007f1df51557b8,0x00007f1df51557f8] = 64
 scopes pcs     [0x00007f1df51557f8,0x00007f1df5155848] = 80
 dependencies   [0x00007f1df5155848,0x00007f1df5155850] = 8
scopes:
ScopeDesc(pc=0x00007f1df515573a offset=1a):
   TestMultiSFO::test@-1 (line 12)
ScopeDesc(pc=0x00007f1df515575c offset=3c):
   TestMultiSFO::test@28 (line 19)
   Locals
    - l0: empty
    - l1: empty
    - l2: empty
    - l3: merge: ID=14
    - l4: obj: ID=15, is_root=1, N.Fields=2, klass: TestMultiSFO$Point 
Fields: stack[12], stack[8]

   Objects
    - 0: merge: ID=14, selector="reg rbp [10],int", merge_pointer="nullptr", candidate objs=[15, 16]
    - 1: obj: ID=15, is_root=1, N.Fields=2, klass: TestMultiSFO$Point 
Fields: stack[12], stack[8]
    - 2: obj: ID=16, is_root=0, N.Fields=2, klass: TestMultiSFO$Point 
Fields: stack[8], stack[12]
ScopeDesc(pc=0x00007f1df5155778 offset=58):
   TestMultiSFO::test@28 (line 19)
```

Reply:
> Why is it limited to non-product builds? It's valuable irrespective of build flavor.

This is because `print_on` in `AnyObj` is only defined in non-product builds. I based implementation of `ObjectMergeValue::print_on` on `ObjectValue::print_on`. In `ObjectValue::print_on` fields aren't printed in product builds.

> Any particular reason for that? You could add is_object_merge case in ObjectValue::print_oninstead and extendObjectValue::print_fields_onto coverObjectMergeValue case.

I'll do that then.

> Also, formatting is broken.

Can you please share an example? If you mean the tabs on lines 303/304/306/307 I added those because I thought would make the code easier to read, but if you want I can definitely remove that.

Reply:
>  If you mean the tabs on lines 303/304/306/307

Yes, it confused me. As an alternative, you could put selector and merge_pointer-related statements on the same line, but I'm not sure how much it improves readability:
```
  st->print(", selector=\""); _selector->print_on(st); st->print("\"");
```

Reply:
A couple of suggestions about the output:
* `merge`: it's clearer to call it `merge_obj`
* `obj` vs `merge` output: obj output is duplicated in ScopeDesc entries and Objects sections; before it was a short version printed in Locals/Expressions and all the details were included in Objects; I like to see field locations in the short version, but including everything looks way too much IMO;
* it makes sense to include selector and merge_pointer info in short version, but `is_root` can be omitted

Reply:
Thanks @iwanowww . Does the output below look good to you? It prints ObjectValue in the same format as it was before this PR and only print details of the merge in the "Objects" section. Is there other output section that you think needs to be adjusted?

```
Compiled method (c2)     436   24             TestMultiSFO::test (48 bytes)
 total in heap  [0x00007f1df5155590,0x00007f1df5155850] = 704
 relocation     [0x00007f1df5155700,0x00007f1df5155718] = 24
 main code      [0x00007f1df5155720,0x00007f1df5155788] = 104
 stub code      [0x00007f1df5155788,0x00007f1df51557a0] = 24
 oops           [0x00007f1df51557a0,0x00007f1df51557b0] = 16
 metadata       [0x00007f1df51557b0,0x00007f1df51557b8] = 8
 scopes data    [0x00007f1df51557b8,0x00007f1df51557f8] = 64
 scopes pcs     [0x00007f1df51557f8,0x00007f1df5155848] = 80
 dependencies   [0x00007f1df5155848,0x00007f1df5155850] = 8
scopes:
ScopeDesc(pc=0x00007f1df515573a offset=1a):
   TestMultiSFO::test@-1 (line 12)
ScopeDesc(pc=0x00007f1df515575c offset=3c):
   TestMultiSFO::test@28 (line 19)
   Locals
    - l0: empty
    - l1: empty
    - l2: empty
    - l3: merge_obj[14]
    - l4: obj[15]

   Objects
    - merge_obj[14], selector="reg rbp [10],int", merge_pointer="nullptr", candidate_objs=[15, 16]
    - obj[15], is_root=1, klass: TestMultiSFO$Point 
	  Fields: stack[12], stack[8]
    - obj[16], is_root=0, klass: TestMultiSFO$Point 
	  Fields: stack[8], stack[12]
```

Reply:
Thanks, it looks much better now (except the position in Objects array is missing). 

It makes sense to mention `is_root` for merge_obj case even though it's always equals to '1`. Also, make merge_pointer optional and omit it when its value is null.

BTW instead of printing `is_root=0/1`, you can introduce a more compact notation for and mark relevant lines with a single symbol:
```
    - 0: R merge_obj[14], selector="reg rbp [10],int", merge_pointer="nullptr", candidate_objs=[15, 16]
    - 1: R obj[15], is_root=1, klass: TestMultiSFO$Point 
 	        Fields: stack[12], stack[8]
    - 2:   obj[16], is_root=0, klass: TestMultiSFO$Point 
	        Fields: stack[8], stack[12]
```

Reply:
Sounds good. I'll make the changes and push them asap.